### PR TITLE
Speed improvement to HGCDigitizer

### PR DESCRIPTION
**Pull request by lgray**
This PR implements smarter rehashing within the HGCDigitizer classes yielding a factor 7 improvement in speed of that module in high pileup. There is also a scheme to keep the hash table size under control using a running average of the occupancy, this allows us to maintain a slim memory profile and small total number of rehashes.

Results below are with 200PU in D4 using CMSSW_9_0_X_2016-12-11-110.

igprof output before patch:
```
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 
            0.0  .........       0.02 / 2'873.81     <spontaneous> [1]
           65.4  .........   1'880.77 / 2'121.26     edm::MixingModule::pileAllWorkers(edm::EventPrincipal const&, edm::ModuleCallingContext const*, int, int, int&, edm::EventSetup const&, edm::StreamID const&) [28]
[29]       65.4   1'880.79       0.13 / 1'880.66   edm::MixingModule::accumulateEvent(PileUpEventPrincipal const&, edm::EventSetup const&, edm::StreamID const&)
           42.2  .........   1'213.16 / 1'213.16     HGCDigitizer::accumulate(PileUpEventPrincipal const&, edm::EventSetup const&, CLHEP::HepRandomEngine*) [30]
            9.7  .........     279.27 / 279.27       void cms::Phase2TrackerDigitizer::accumulate_local<PileUpEventPrincipal>(PileUpEventPrincipal const&, edm::EventSetup const&) [43]
            5.4  .........     154.53 / 154.53       EcalDigiProducer::accumulate(PileUpEventPrincipal const&, edm::EventSetup const&, edm::StreamID const&) [96]
            4.7  .........     134.16 / 134.16       TrackingTruthAccumulator::accumulate(PileUpEventPrincipal const&, edm::EventSetup const&, edm::StreamID const&) [105]
            1.8  .........      52.12 / 52.14        cms::PileupVertexAccumulator::accumulate(PileUpEventPrincipal const&, edm::EventSetup const&, edm::StreamID const&) [163]
            1.1  .........      30.56 / 30.56        HcalDigitizer::accumulate(PileUpEventPrincipal const&, edm::EventSetup const&, CLHEP::HepRandomEngine*) [616]
            0.5  .........      14.09 / 14.09        CaloTruthAccumulator::accumulate(PileUpEventPrincipal const&, edm::EventSetup const&, edm::StreamID const&) [1008]
            0.1  .........       1.98 / 1.98         EcalTimeDigiProducer::accumulate(PileUpEventPrincipal const&, edm::EventSetup const&, edm::StreamID const&) [1427]
            0.0  .........       0.68 / 0.68         SiStripDigitizer::accumulate(PileUpEventPrincipal const&, edm::EventSetup const&, edm::StreamID const&) [1940]
            0.0  .........       0.05 / 0.05         HGCDigiProducer::accumulate(PileUpEventPrincipal const&, edm::EventSetup const&, edm::StreamID const&) [3987]
            0.0  .........       0.04 / 0.04         HcalDigiProducer::accumulate(PileUpEventPrincipal const&, edm::EventSetup const&, edm::StreamID const&) [4145]
            0.0  .........       0.02 / 0.02         cms::Phase2TrackerDigitizer::accumulate(PileUpEventPrincipal const&, edm::EventSetup const&, edm::StreamID const&) [5949]

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 
```

igprof output with patch:
```
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 
            0.0  .........       0.02 / 1'773.66     <spontaneous> [1]
           46.3  .........     821.99 / 1'062.78     edm::MixingModule::pileAllWorkers(edm::EventPrincipal const&, edm::ModuleCallingContext const*, int, int, int&, edm::EventSetup const&, edm::StreamID const&) [28]
[29]       46.3     822.01       0.11 / 821.90     edm::MixingModule::accumulateEvent(PileUpEventPrincipal const&, edm::EventSetup const&, edm::StreamID const&)
           15.5  .........     274.75 / 274.75       void cms::Phase2TrackerDigitizer::accumulate_local<PileUpEventPrincipal>(PileUpEventPrincipal const&, edm::EventSetup const&) [40]
            9.6  .........     170.05 / 170.05       HGCDigitizer::accumulate(PileUpEventPrincipal const&, edm::EventSetup const&, CLHEP::HepRandomEngine*) [86]
            8.6  .........     152.25 / 152.25       EcalDigiProducer::accumulate(PileUpEventPrincipal const&, edm::EventSetup const&, edm::StreamID const&) [95]
            7.3  .........     130.35 / 130.35       TrackingTruthAccumulator::accumulate(PileUpEventPrincipal const&, edm::EventSetup const&, edm::StreamID const&) [105]
            2.8  .........      50.13 / 50.19        cms::PileupVertexAccumulator::accumulate(PileUpEventPrincipal const&, edm::EventSetup const&, edm::StreamID const&) [162]
            1.6  .........      28.95 / 28.95        HcalDigitizer::accumulate(PileUpEventPrincipal const&, edm::EventSetup const&, CLHEP::HepRandomEngine*) [637]
            0.8  .........      13.32 / 13.32        CaloTruthAccumulator::accumulate(PileUpEventPrincipal const&, edm::EventSetup const&, edm::StreamID const&) [1011]
            0.1  .........       1.45 / 1.45         EcalTimeDigiProducer::accumulate(PileUpEventPrincipal const&, edm::EventSetup const&, edm::StreamID const&) [1519]
            0.0  .........       0.56 / 0.56         SiStripDigitizer::accumulate(PileUpEventPrincipal const&, edm::EventSetup const&, edm::StreamID const&) [2010]
            0.0  .........       0.03 / 0.03         HcalDigiProducer::accumulate(PileUpEventPrincipal const&, edm::EventSetup const&, edm::StreamID const&) [5124]
            0.0  .........       0.02 / 0.02         HGCDigiProducer::accumulate(PileUpEventPrincipal const&, edm::EventSetup const&, edm::StreamID const&) [5406]
            0.0  .........       0.02 / 0.02         cms::Phase2TrackerDigitizer::accumulate(PileUpEventPrincipal const&, edm::EventSetup const&, edm::StreamID const&) [5409]

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 
```

virtual and rss before patch:
```
MemoryReport> Peak virtual size 5676.52 Mbytes
 Key events increasing vsize: 
[2] run: 1 lumi: 1 event: 5  vsize = 3972.17 deltaVsize = 160 rss = 2923.42 delta = 425.906
[6] run: 1 lumi: 1 event: 20  vsize = 4556.52 deltaVsize = 224 rss = 3304.84 delta = 381.422
[8] run: 1 lumi: 1 event: 1  vsize = 5324.52 deltaVsize = 672 rss = 3686.03 delta = 381.188
[9] run: 1 lumi: 1 event: 18  vsize = 5676.52 deltaVsize = 352 rss = 3875.34 delta = 189.309
[0] run: 0 lumi: 0 event: 0  vsize = 0 deltaVsize = 0 rss = 0 delta = 0
[8] run: 1 lumi: 1 event: 1  vsize = 5324.52 deltaVsize = 672 rss = 3686.03 delta = 276.277
[10] run: 1 lumi: 1 event: 37  vsize = 5676.52 deltaVsize = 0 rss = 3687.72 delta = -187.617
[9] run: 1 lumi: 1 event: 18  vsize = 5676.52 deltaVsize = 352 rss = 3875.34 delta = 189.309
```

virtual and rss after patch:
```
MemoryReport> Peak virtual size 5646.11 Mbytes
 Key events increasing vsize: 
[2] run: 1 lumi: 1 event: 5  vsize = 3973.76 deltaVsize = 160 rss = 2930.69 delta = 442.617
[5] run: 1 lumi: 1 event: 7  vsize = 4398.11 deltaVsize = 168 rss = 3194.73 delta = 264.039
[8] run: 1 lumi: 1 event: 1  vsize = 5262.11 deltaVsize = 672 rss = 3718.68 delta = 523.949
[9] run: 1 lumi: 1 event: 18  vsize = 5454.11 deltaVsize = 192 rss = 3851.01 delta = 132.328
[10] run: 1 lumi: 1 event: 37  vsize = 5646.11 deltaVsize = 192 rss = 3870.63 delta = 19.6211
[8] run: 1 lumi: 1 event: 1  vsize = 5262.11 deltaVsize = 672 rss = 3718.68 delta = 317.082
[9] run: 1 lumi: 1 event: 18  vsize = 5454.11 deltaVsize = 192 rss = 3851.01 delta = 132.328
[10] run: 1 lumi: 1 event: 37  vsize = 5646.11 deltaVsize = 192 rss = 3870.63 delta = 19.6211
```

### CODE REVIEW

**Comment by lgray** [comment]
@cmsbuild please test

**Comment by lgray** [comment]
--More detailed igprof performance information--

module before patch:
```
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 
Rank    % total       Self       Self / Children   Function
            0.0  .........       0.38 / 0.40         HGCDigitizer::accumulate(edm::Event const&, edm::EventSetup const&, CLHEP::HepRandomEngine*) [2244]
           41.6  .........   1'194.23 / 1'213.16     HGCDigitizer::accumulate(PileUpEventPrincipal const&, edm::EventSetup const&, CLHEP::HepRandomEngine*) [30]
[31]       41.6   1'194.60      19.34 / 1'175.26   void HGCDigitizer::accumulate<HGCalGeometry>(edm::Handle<std::vector<PCaloHit, std::allocator<PCaloHit> > > const&, int, HGCalGeometry const*, CLHEP::HepRandomEngine*)
           35.1  .........   1'007.54 / 1'015.85     std::_Hashtable<unsigned int, std::pair<unsigned int const, hgc_digi::HGCCellInfo>, std::allocator<std::pair<unsigned int const, hgc_digi::HGCCellInfo> >, std::__detail::_Select1st, std::equal_to<unsigned int>, std::hash<unsigned int>, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<false, false, true> >::_M_rehash(unsigned long, unsigned long const&) [32]
            2.5  .........      72.50 / 75.04        std::pair<std::__detail::_Node_iterator<std::pair<unsigned int const, hgc_digi::HGCCellInfo>, false, false>, bool> std::_Hashtable<unsigned int, std::pair<unsigned int const, hgc_digi::HGCCellInfo>, std::allocator<std::pair<unsigned int const, hgc_digi::HGCCellInfo> >, std::__detail::_Select1st, std::equal_to<unsigned int>, std::hash<unsigned int>, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<false, false, true> >::_M_emplace<unsigned int const&, hgc_digi::HGCCellInfo>(std::integral_constant<bool, true>, unsigned int const&, hgc_digi::HGCCellInfo&&) [129]
            1.9  .........      53.90 / 56.93        std::_Hashtable<DetId, DetId, std::allocator<DetId>, std::__detail::_Identity, std::equal_to<DetId>, std::hash<DetId>, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<true, true, true> >::count(DetId const&) const [153]
            0.6  .........      16.52 / 16.52        HGCalGeometry::getPosition(DetId const&) const [986]
            0.4  .........      11.21 / 11.33        void std::__introsort_loop<__gnu_cxx::__normal_iterator<std::tuple<int, unsigned int, float>*, std::vector<std::tuple<int, unsigned int, float>, std::allocator<std::tuple<int, unsigned int, float> > > >, long, __gnu_cxx::__ops::_Iter_comp_iter<bool (*)(std::tuple<int, unsigned int, float> const&, std::tuple<int, unsigned int, float> const&)> >(__gnu_cxx::__normal_iterator<std::tuple<int, unsigned int, float>*, std::vector<std::tuple<int, unsigned int, float>, std::allocator<std::tuple<int, unsigned int, float> > > >, __gnu_cxx::__normal_iterator<std::tuple<int, unsigned int, float>*, std::vector<std::tuple<int, unsigned int, float>, std::allocator<std::tuple<int, unsigned int, float> > > >, long, __gnu_cxx::__ops::_Iter_comp_iter<bool (*)(std::tuple<int, unsigned int, float> const&, std::tuple<int, unsigned int, float> const&)>) [1038]
            0.2  .........       5.39 / 5.50         void std::__unguarded_linear_insert<__gnu_cxx::__normal_iterator<std::tuple<int, unsigned int, float>*, std::vector<std::tuple<int, unsigned int, float>, std::allocator<std::tuple<int, unsigned int, float> > > >, __gnu_cxx::__ops::_Val_comp_iter<bool (*)(std::tuple<int, unsigned int, float> const&, std::tuple<int, unsigned int, float> const&)> >(__gnu_cxx::__normal_iterator<std::tuple<int, unsigned int, float>*, std::vector<std::tuple<int, unsigned int, float>, std::allocator<std::tuple<int, unsigned int, float> > > >, __gnu_cxx::__ops::_Val_comp_iter<bool (*)(std::tuple<int, unsigned int, float> const&, std::tuple<int, unsigned int, float> const&)>) [1182]
            0.1  .........       3.15 / 3.52         HGCalDDDConstants::simToReco(int, int, int, bool) const [1266]
            0.1  .........       1.63 / 1.67         floor [1492]
            0.0  .........       1.33 / 1.40         HGCalTestNumbering::unpackHexagonIndex(unsigned int const&, int&, int&, int&, int&, int&, int&) [1566]
            0.0  .........       0.58 / 0.60         void std::vector<std::tuple<int, unsigned int, float>, std::allocator<std::tuple<int, unsigned int, float> > >::emplace_back<int&, unsigned int, float>(int&, unsigned int&&, float&&) [2006]
            0.0  .........       0.50 / 20.89        _init [966]
            0.0  .........       0.50 / 0.58         HGCalDetId::HGCalDetId(ForwardSubdetector, int, int, int, int, int) [2029]
            0.0  .........       0.28 / 0.34         edm::HandleBase::productStorage() const [2333]
            0.0  .........       0.08 / 0.13         std::vector<std::tuple<int, unsigned int, float>, std::allocator<std::tuple<int, unsigned int, float> > >::reserve(unsigned long) [3031]
            0.0  .........       0.05 / 0.06         void std::__insertion_sort<__gnu_cxx::__normal_iterator<std::tuple<int, unsigned int, float>*, std::vector<std::tuple<int, unsigned int, float>, std::allocator<std::tuple<int, unsigned int, float> > > >, __gnu_cxx::__ops::_Iter_comp_iter<bool (*)(std::tuple<int, unsigned int, float> const&, std::tuple<int, unsigned int, float> const&)> >(__gnu_cxx::__normal_iterator<std::tuple<int, unsigned int, float>*, std::vector<std::tuple<int, unsigned int, float>, std::allocator<std::tuple<int, unsigned int, float> > > >, __gnu_cxx::__normal_iterator<std::tuple<int, unsigned int, float>*, std::vector<std::tuple<int, unsigned int, float>, std::allocator<std::tuple<int, unsigned int, float> > > >, __gnu_cxx::__ops::_Iter_comp_iter<bool (*)(std::tuple<int, unsigned int, float> const&, std::tuple<int, unsigned int, float> const&)>) [3739]
            0.0  .........       0.04 / 0.07         std::__detail::_Prime_rehash_policy::_M_next_bkt(unsigned long) const [3523]
            0.0  .........       0.02 / 1.39         je_huge_dalloc [1568]
            0.0  .........       0.01 / 0.22         je_arena_dalloc_large [2637]
            0.0  .........       0.01 / 0.01         ceilf [6719]
            0.0  .........       0.01 / 0.75         _dl_runtime_resolve [1885]
            0.0  .........       0.01 / 0.01         ceil [8494]

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 
```

module with patch:
```
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 
            0.0  .........       0.33 / 0.35         HGCDigitizer::accumulate(edm::Event const&, edm::EventSetup const&, CLHEP::HepRandomEngine*) [2282]
            9.1  .........     161.80 / 170.05       HGCDigitizer::accumulate(PileUpEventPrincipal const&, edm::EventSetup const&, CLHEP::HepRandomEngine*) [86]
[92]        9.1     162.13      18.48 / 143.65     void HGCDigitizer::accumulate<HGCalGeometry>(edm::Handle<std::vector<PCaloHit, std::allocator<PCaloHit> > > const&, int, HGCalGeometry const*, CLHEP::HepRandomEngine*)
            3.1  .........      55.55 / 57.10        std::pair<std::__detail::_Node_iterator<std::pair<unsigned int const, hgc_digi::HGCCellInfo>, false, false>, bool> std::_Hashtable<unsigned int, std::pair<unsigned int const, hgc_digi::HGCCellInfo>, std::allocator<std::pair<unsigned int const, hgc_digi::HGCCellInfo> >, std::__detail::_Select1st, std::equal_to<unsigned int>, std::hash<unsigned int>, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<false, false, true> >::_M_emplace<unsigned int const&, hgc_digi::HGCCellInfo>(std::integral_constant<bool, true>, unsigned int const&, hgc_digi::HGCCellInfo&&) [150]
            2.8  .........      48.80 / 51.44        std::_Hashtable<DetId, DetId, std::allocator<DetId>, std::__detail::_Identity, std::equal_to<DetId>, std::hash<DetId>, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<true, true, true> >::count(DetId const&) const [159]
            0.9  .........      16.22 / 16.22        HGCalGeometry::getPosition(DetId const&) const [983]
            0.6  .........      10.40 / 10.47        void std::__introsort_loop<__gnu_cxx::__normal_iterator<std::tuple<int, unsigned int, float>*, std::vector<std::tuple<int, unsigned int, float>, std::allocator<std::tuple<int, unsigned int, float> > > >, long, __gnu_cxx::__ops::_Iter_comp_iter<bool (*)(std::tuple<int, unsigned int, float> const&, std::tuple<int, unsigned int, float> const&)> >(__gnu_cxx::__normal_iterator<std::tuple<int, unsigned int, float>*, std::vector<std::tuple<int, unsigned int, float>, std::allocator<std::tuple<int, unsigned int, float> > > >, __gnu_cxx::__normal_iterator<std::tuple<int, unsigned int, float>*, std::vector<std::tuple<int, unsigned int, float>, std::allocator<std::tuple<int, unsigned int, float> > > >, long, __gnu_cxx::__ops::_Iter_comp_iter<bool (*)(std::tuple<int, unsigned int, float> const&, std::tuple<int, unsigned int, float> const&)>) [1043]
            0.3  .........       5.01 / 5.11         void std::__unguarded_linear_insert<__gnu_cxx::__normal_iterator<std::tuple<int, unsigned int, float>*, std::vector<std::tuple<int, unsigned int, float>, std::allocator<std::tuple<int, unsigned int, float> > > >, __gnu_cxx::__ops::_Val_comp_iter<bool (*)(std::tuple<int, unsigned int, float> const&, std::tuple<int, unsigned int, float> const&)> >(__gnu_cxx::__normal_iterator<std::tuple<int, unsigned int, float>*, std::vector<std::tuple<int, unsigned int, float>, std::allocator<std::tuple<int, unsigned int, float> > > >, __gnu_cxx::__ops::_Val_comp_iter<bool (*)(std::tuple<int, unsigned int, float> const&, std::tuple<int, unsigned int, float> const&)>) [1187]
            0.2  .........       2.91 / 3.19         HGCalDDDConstants::simToReco(int, int, int, bool) const [1275]
            0.1  .........       1.76 / 1.82         floor [1428]
            0.1  .........       1.01 / 1.09         HGCalTestNumbering::unpackHexagonIndex(unsigned int const&, int&, int&, int&, int&, int&, int&) [1644]
            0.0  .........       0.61 / 0.61         void std::vector<std::tuple<int, unsigned int, float>, std::allocator<std::tuple<int, unsigned int, float> > >::emplace_back<int&, unsigned int, float>(int&, unsigned int&&, float&&) [1938]
            0.0  .........       0.51 / 0.55         HGCalDetId::HGCalDetId(ForwardSubdetector, int, int, int, int, int) [2029]
            0.0  .........       0.39 / 20.52        _init [964]
            0.0  .........       0.31 / 0.38         edm::HandleBase::productStorage() const [2241]
            0.0  .........       0.07 / 0.08         void std::__insertion_sort<__gnu_cxx::__normal_iterator<std::tuple<int, unsigned int, float>*, std::vector<std::tuple<int, unsigned int, float>, std::allocator<std::tuple<int, unsigned int, float> > > >, __gnu_cxx::__ops::_Iter_comp_iter<bool (*)(std::tuple<int, unsigned int, float> const&, std::tuple<int, unsigned int, float> const&)> >(__gnu_cxx::__normal_iterator<std::tuple<int, unsigned int, float>*, std::vector<std::tuple<int, unsigned int, float>, std::allocator<std::tuple<int, unsigned int, float> > > >, __gnu_cxx::__normal_iterator<std::tuple<int, unsigned int, float>*, std::vector<std::tuple<int, unsigned int, float>, std::allocator<std::tuple<int, unsigned int, float> > > >, __gnu_cxx::__ops::_Iter_comp_iter<bool (*)(std::tuple<int, unsigned int, float> const&, std::tuple<int, unsigned int, float> const&)>) [3331]
            0.0  .........       0.05 / 0.07         std::vector<std::tuple<int, unsigned int, float>, std::allocator<std::tuple<int, unsigned int, float> > >::reserve(unsigned long) [3480]
            0.0  .........       0.02 / 1.18         je_huge_dalloc [1616]
            0.0  .........       0.02 / 44.95        free [176]
            0.0  .........       0.01 / 0.09         je_arena_dalloc_large [3248]
            0.0  .........       0.01 / 0.67         _dl_runtime_resolve [1885]

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 
```

**Comment by lgray** [comment]
@slava77 FYI

**Comment by lgray** [comment]
@civanch Could you please sign? I see no significant differences compared to the baseline. Thanks!

**Comment by civanch** [comment]
+1

**Comment by davidlange6** [comment]
+1

File: SimCalorimetry/HGCalSimProducers/src/HGCDigitizer.cc [hunk 1]
```diff
@@ -142,7 +162,30 @@ void HGCDigitizer::finalizeEvent(edm::Event& e, edm::EventSetup const& es, CLHEP
...
+    break;
+  case ForwardSubdetector::HGCHEB:
+    idx = 2;
+    break;
+  default:
+    break;
+  }
+  // release memory for unfilled parts of hash table
+  if( validIds_.size()*averageOccupancies_[idx] > simHitAccumulator_->size() ) {
+    simHitAccumulator_->reserve(simHitAccumulator_->size());
```
Comment (makortel) [inline SimCalorimetry/HGCalSimProducers/src/HGCDigitizer.cc]: Sorry to raise an old PR, but what is the aim of this line? I'd expect this to be a no-op.

(and what would be the better way to ask random questions on random pieces of code?)

**Pull request merged**